File: bls-server/src/shared/utils/snowflake.ts
//雪花算法生成ID

const EPOCH = 1704067200000n; // 2024-01-01 00:00:00 UTC
const WORKER_ID_BITS = 5n;
const DATACENTER_ID_BITS = 5n;
const SEQUENCE_BITS = 12n;

const MAX_WORKER_ID = (1n << WORKER_ID_BITS) - 1n;
const MAX_DATACENTER_ID = (1n << DATACENTER_ID_BITS) - 1n;
const SEQUENCE_MASK = (1n << SEQUENCE_BITS) - 1n;

const WORKER_ID_SHIFT = SEQUENCE_BITS;
const DATACENTER_ID_SHIFT = SEQUENCE_BITS + WORKER_ID_BITS;
const TIMESTAMP_SHIFT = SEQUENCE_BITS + WORKER_ID_BITS + DATACENTER_ID_BITS;

let sequence = 0n;
let lastTimestamp = -1n;

const workerId = 1n;
const datacenterId = 1n;

function currentTimestamp(): bigint {
  return BigInt(Date.now());
}

function waitNextMillis(lastTs: bigint): bigint {
  let ts = currentTimestamp();
  while (ts <= lastTs) {
    ts = currentTimestamp();
  }
  return ts;
}

export function generateSnowflakeId(): string {
  let timestamp = currentTimestamp();

  if (timestamp < lastTimestamp) {
    throw new Error('系统时钟回拨，无法生成雪花ID');
  }

  if (timestamp === lastTimestamp) {
    sequence = (sequence + 1n) & SEQUENCE_MASK;
    if (sequence === 0n) {
      timestamp = waitNextMillis(lastTimestamp);
    }
  } else {
    sequence = 0n;
  }

  lastTimestamp = timestamp;

  const id = ((timestamp - EPOCH) << TIMESTAMP_SHIFT)
    | (datacenterId << DATACENTER_ID_SHIFT)
    | (workerId << WORKER_ID_SHIFT)
    | sequence;

  return id.toString();
}

export function generatePlatformTenantId(): string {
  return '000000';
}

File: bls-rust-server/src/utils/snowflake.rs
use std::sync::atomic::{AtomicI64, AtomicU64, Ordering};
use std::time::{SystemTime, UNIX_EPOCH};

const EPOCH_MS: u64 = 1_704_067_200_000;
const SEQUENCE_BITS: u64 = 12;
const WORKER_ID_BITS: u64 = 5;
const DATACENTER_ID_BITS: u64 = 5;
const SEQUENCE_MASK: u64 = (1 << SEQUENCE_BITS) - 1;

pub struct SnowflakeGenerator {
    worker_id: u64,
    datacenter_id: u64,
    last_timestamp: AtomicI64,
    sequence: AtomicU64,
}

impl SnowflakeGenerator {
    pub fn new(worker_id: u64, datacenter_id: u64) -> anyhow::Result<Self> {
        if worker_id >= (1 << WORKER_ID_BITS) || datacenter_id >= (1 << DATACENTER_ID_BITS) {
            anyhow::bail!("invalid snowflake worker/datacenter id");
        }
        Ok(Self {
            worker_id,
            datacenter_id,
            last_timestamp: AtomicI64::new(-1),
            sequence: AtomicU64::new(0),
        })
    }

    pub fn next_id(&self) -> anyhow::Result<String> {
        let mut timestamp = now_ms();
        let mut last = self.last_timestamp.load(Ordering::SeqCst);
        if timestamp < last {
            anyhow::bail!("clock moved backwards");
        }
        let sequence = if timestamp == last {
            let seq = self.sequence.fetch_add(1, Ordering::SeqCst) & SEQUENCE_MASK;
            if seq == 0 {
                timestamp = wait_next_ms(last);
                self.last_timestamp.store(timestamp, Ordering::SeqCst);
            }
            seq
        } else {
            self.sequence.store(0, Ordering::SeqCst);
            self.last_timestamp.store(timestamp, Ordering::SeqCst);
            0
        };

        let id = ((timestamp as u64 - EPOCH_MS)
            << (SEQUENCE_BITS + WORKER_ID_BITS + DATACENTER_ID_BITS))
            | (self.datacenter_id << (SEQUENCE_BITS + WORKER_ID_BITS))
            | (self.worker_id << SEQUENCE_BITS)
            | sequence;
        Ok(id.to_string())
    }
}

fn now_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}

fn wait_next_ms(last: i64) -> i64 {
    let mut ts = now_ms();
    while ts <= last {
        ts = now_ms();
    }
    ts
}
